app/src/main/java/nl/eduvpn/app/service/WireGuardService.kt

```diff
@@ -0,0 +1,150 @@
+import com.wireguard.config.Interface
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import nl.eduvpn.app.livedata.ByteCount
+import nl.eduvpn.app.livedata.IPs
+import nl.eduvpn.app.utils.Log
+import nl.eduvpn.app.utils.WireGuardTunnel
+
+/**
+ * Service responsible for managing the WireGuard profiles and the connection.
+ */
+class WireGuardService(private val context: Context, timer: LiveData<Unit>) :
+    VPNService() {
+
+    private val backend = GoBackend(context)
+
+    private var errorString: String? = null
+
+    // Stores the current VPN status.
+    private var connectionStatus = VPNStatus.DISCONNECTED
+
+    private val TAG = this::class.java.name
+
+    private val tunnel = WireGuardTunnel("eduVPN WireGuard tunnel") { newTunnelState ->
+        setConnectionStatus(tunnelStateToStatus(newTunnelState))
+    }
+
+    override fun getStatus(): VPNStatus {
+        return connectionStatus
+    }
+
+    override fun startForeground(id: Int, notification: Notification) {
+        // Not necessary for WireGuard
+    }
+
+    private fun setConnectionStatus(status: VPNStatus) {
+        connectionStatus = status
+        postValue(status)
+    }
+
+    /**
+     * Ask user for VPN permission
+     */
+    private fun authorizeVPN(activity: Activity) {
+        val intent = GoBackend.VpnService.prepare(context)
+        if (intent != null) {
+            activity.startActivityForResult(intent, 0)
+        }
+    }
+
+    override val byteCountLiveData: LiveData<ByteCount?> = timer.map { getByteCount() }
+
+    override val ipLiveData: MutableLiveData<IPs> = MutableLiveData()
+
+    private fun getByteCount(): ByteCount {
+        val statistics = backend.getStatistics(tunnel)
+        val bytesIn = statistics.totalRx()
+        val bytesOut = statistics.totalTx()
+        return ByteCount(bytesIn, bytesOut)
+    }
+
+    /**
+     * Connects to the VPN using the config supplied as a parameter.
+     *
+     * @param activity   The current activity, required for providing a context.
+     * @param config  The config to use for connecting.
+     */
+    suspend fun connect(activity: Activity, config: Config) {
+        setConnectionStatus(VPNStatus.CONNECTING)
+
+        ipLiveData.postValue(getIPs(config.`interface`))
+
+        withContext(Dispatchers.IO) {
+            try {
+                backend.setState(tunnel, Tunnel.State.UP, config)
+            } catch (ex: BackendException) {
+                if (ex.reason == BackendException.Reason.VPN_NOT_AUTHORIZED) {
+                    withContext(Dispatchers.Main) {
+                        authorizeVPN(activity)
+                    }
+                    setConnectionStatus(VPNStatus.DISCONNECTED)
+                } else {
+                    fail(ex.toString())
+                }
+            }
+        }
+    }
+
+    private fun fail(errorString: String) {
+        setConnectionStatus(VPNStatus.FAILED)
+        this.errorString = errorString
```

Comment:
You should set the error string before notifying the subscribers about the failure